workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt

```diff
@@ -0,0 +1,180 @@
+package com.squareup.workflow1.ui
+
+import android.view.View
+import android.view.View.OnAttachStateChangeListener
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.VisibleForTesting.PRIVATE
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.Event
+import androidx.lifecycle.Lifecycle.State.DESTROYED
+import androidx.lifecycle.Lifecycle.State.RESUMED
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+import androidx.lifecycle.LifecycleRegistry.createUnsafe
+import androidx.lifecycle.ViewTreeLifecycleOwner
+
+/**
+ * An extension of [LifecycleOwner] that is always owned by a [View], is logically a child lifecycle
+ * of the next-nearest [ViewTreeLifecycleOwner] above it (it mirrors its parent's lifecycle until
+ * it's destroyed), and can be [asked to destroy][destroyOnDetach] itself early.
+ *
+ * This type is meant to help integrate with [ViewTreeLifecycleOwner] by allowing the creation of a
+ * tree of [LifecycleOwner]s that mirrors the view tree.
+ *
+ * Custom container views that use [ViewRegistry.buildView] to create their children _must_ ensure
+ * they call [destroyOnDetach] on the outgoing view before they replace children with new views.
```

Comment:
And it doesn't make sense for us to wrap this swap (enforcing `destroyOnDetach`) in a base container type?

Reply:
How would we do that? The containers all have different logic around this code. E.g. compare `BackStackContainer` and `ModalContainer`.

Reply:
It probably makes sense to do this automatically in Compose, since I expect container implementations will all look a lot more uniform and Compose actually has a stronger concept of lifetime than classic android views do.